File: R/acdx.R
acdx <- function(
  ac,           # aggregated cell object
  X=NULL,       # design matrix for mean parameters
  Gi=NULL,      # group intercepts for dispersion parameters

  n_boot=1,     # number of bootstrap resampling
  id_boot=NULL, # block bootstrap id (default: sample id)
  seed_boot=0,  # random number seed
  
  pbulk = FALSE,    # "pseudo-bulk" mode
  mean2sum = TRUE,  # convert y from mean to sum if pbulk==TRUE

  norm_method = 1, # normalization methoda: 0 none, 1 per celltype, 2 global

  y0=NULL,      # small count to be added, 0.5*min(y|y > 0)
  verbose=0
  )
{

  n_sample <- nrow(ac$N)
  n_ctype <- ncol(ac$N)
  n_gene <- dim(ac$y)[3]
  if(!is.null(X))
    { n_coef <- ncol(X); Xt <- t(X) }
  else
    { n_coef <- 0; Xt <- NULL }

  if(is.null(Gi)) Gi <- rep(0,n_sample)
  Gi_names <- unique(Gi)
  Gi <- as.integer(factor(Gi,levels=Gi_names))-1
  n_disp <- length(unique(Gi))

  if(is.null(id_boot)) id_boot <- 0:(n_sample-1)
  id_boot <- as.integer(factor(id_boot,levels=unique(id_boot)))-1
  n_bid <- length(unique(id_boot))
 
  set.seed(seed_boot)
  B <- sapply(1:n_boot,function(u)sample.int(n_bid,n_bid,replace=TRUE)-1)
  B[,1] <- 0:(n_bid-1)

  if(pbulk) # pseudo-bulk mode
    {
    if(mean2sum)
      for(i in 1:dim(ac$y)[4])
        ac$y[1,,,i] <- ac$y[1,,,i] * ac$N[,i]
    family <- 1  # negative binomial model
    }
  else
    {
    family <- 0 # REGa model
    }

  if(is.null(y0))
    y0 <- 0.5*min(ac$y[1,,,][ ac$y[1,,,] > 0],na.rm=T)

  if( !(norm_method == 1 || norm_method == 2) )
    norm_method <- 0

  r <- .C("malm_acdx",
    dim=as.integer(c(n_sample,n_gene,n_coef,n_disp,n_ctype,n_bid,n_boot)),
    Y=as.double(ac$y),
    N=as.double(ac$N),
    Xt=as.double(Xt),
    Gi=as.integer(Gi),
    id_boot=as.integer(id_boot),
    B=as.integer(B),
    beta0=double(n_gene*n_ctype*n_boot),
    phi0=double(n_gene*n_ctype*n_boot),
    alpha=double(n_sample*n_ctype*n_boot),
    beta=double(n_coef*n_gene*n_ctype*n_boot),
    phi=double(n_disp*n_gene*n_ctype*n_boot),
    y0=as.double(y0),
    norm_method=as.integer(norm_method),
    iopt11=as.integer(c(family,verbose,acdx.ctrl$iter_max11)),
    dopt11=as.double(c(acdx.ctrl$epsilon11)),
    iopt=as.integer(c(family,verbose,acdx.ctrl$iter_max,acdx.ctrl$bt_iter_max)),
    dopt=as.double(c(acdx.ctrl$epsilon,acdx.ctrl$bt_tol,acdx.ctrl$bt_step)),
    NAOK=TRUE,
    PACKAGE="acdx"
    )
  r$Y <- NULL
  r$N <- NULL
  sample_names <- rownames(ac$N)
  ctype_names <- colnames(ac$N)
  gene_names <- dimnames(ac$y)[[3]]

  dim(r$beta0) <- c(n_gene,n_ctype,n_boot)
  dim(r$phi0) <- c(n_gene,n_ctype,n_boot)
  dimnames(r$beta0) <- dimnames(r$phi0) <- list(gene_names,ctype_names,NULL)
  dim(r$alpha) <- c(n_sample,n_ctype,n_boot)
  dimnames(r$alpha) <- list(sample_names,ctype_names,NULL)

  if(!is.null(X))
    {
    dim(r$beta) <- c(n_coef,n_gene,n_ctype,n_boot)
    dimnames(r$beta) <- list(colnames(X),gene_names,ctype_names,NULL)
    dim(r$phi) <- c(n_disp,n_gene,n_ctype,n_boot)
    dimnames(r$phi) <- list( Gi_names,gene_names,ctype_names,NULL)
    }
  else
    {
    r$beta <- r$phi <- NULL
    }

  if( norm_method != 1 && norm_method != 2 )
    r$beta0 <- r$phi0 <- NULL

  r$seed_boot <- seed_boot
  dim(r$B) <- c(n_bid,n_boot)
  
  structure(r,class="acdx_fit")
}

# iteration control options
acdx.ctrl <- list(
    iter_max11 = 30, epsilon11 = 1e-4,
    iter_max = 30, epsilon = 1e-4,
    bt_iter_max = 4,
    bt_tol = 1e-4,
    bt_step = 0.5)

File: src/acdx.c
/*
 * malm_acdx.c
 *
 *  - input:
 *    - aggregate data from greg (sample x gene x ctype)
 *    - design matrix x
 *    - bootstrap resamples
 *
 *  - repeat for each bootstrap instances (first one is the original)
 *    - repeat for each cell-type:
 *      - normalize using malm11: alpha, beta
 *      - malm for each gene, with log(alpha) as offset [optional]
 *
 *  - output
 *    - (alpha, beta, beta0, phi) x ctype x bootstraps
 */

#include <stdlib.h>
#include <stdio.h>
#include <omp.h>

#include "malm.h"

void
malm_acdx (
  const int *dim,     // dimensions
  uv *Y,              // n_sample x n_gene x n_ctype
  double *N,          // n_sample x n_ctype
  const double *Xt,   // n_coef x n_sample
  const int *Gi,      // n_disp
  const int *id_boot,    // n_sample: map sample to bootstrab block
  const int *B,       // n_bid x n_boot: resampled bid
  double *beta0,      // n_gene x n_ctype x n_boot
  double *phi0,       // n_gene x n_ctype x n_boot
  double *alpha,      // n_sample x n_ctype x n_boot
  double *beta,       // n_coef x n_gene x n_ctype x n_boot
  double *phi,        // n_disp x n_gene x n_ctype x n_boot
  const int *norm_method,
  const int *iopt11_,
  const double *dopt11_,
  const int *iopt_,
  const double *dopt_
  )
{
  int n_sample = dim[0];
  int n_gene = dim[1];
  int n_coef = dim[2];
  int n_disp = dim[3];   // dispersion groups
  int n_ctype = dim[4];
  int n_bid = dim[5];   // number of bootstrap blocks
  int n_boot = dim[6];   // number of bootstrap resamples

  // precompute sample weights for bootstrap, which is also
  // cell-type specific
  double *w = (double*)malloc(sizeof(double)
                  * n_sample * n_ctype * n_boot);
  for(int b = 0; b < n_boot; b++ )
    {
    const int *B_b = B + b * n_bid;
    double *w_b = w + b*n_sample*n_ctype;
    for(int k = 0; k < n_ctype; k++ )
      {
      const double *Nk = N + k*n_sample;
      double *w_bk = w_b + k*n_sample;
      for(int i = 0; i < n_sample; i++ )
        {
        w_bk[i] = 0;
        for(int a = 0; a < n_bid; a++ )
          if( id_boot[i] == B_b[a] ) w_bk[i]++;
        w_bk[i] *= ((Nk[i] > 1 ) ? 1 : 0);
        }
      }
    }


  // allocate thread-specific storage
  //
  const int n_thread = omp_get_max_threads();
  uv *E = (uv*)malloc( sizeof(uv) * n_sample * n_gene * n_thread );

  uv *Yg, *Eg;
  if( *norm_method == 2 )
    {
    Eg = (uv*)malloc( sizeof(uv) * n_sample * n_gene * n_ctype * n_thread);
    Yg = (uv*)malloc( sizeof(uv) * n_sample * n_gene * n_ctype );
    for(int k = 0; k < n_ctype; k++ )
      for(int j = 0; j < n_gene; j++ )
        for(int i = 0; i < n_sample; i++ )
          Yg[i + k*n_sample + j*n_sample*n_ctype ] =
             Y[i + j*n_sample + k*n_sample*n_gene ];
    }

  int n_done = 0, chunk=ceil(n_boot/100.0);

  #pragma omp parallel for schedule(runtime)
 for(int b = 0; b < n_boot; b++ )
    {
    int t_id = omp_get_thread_num();
    const double *w_b = w + b * n_sample * n_ctype;
    double *alpha_b = alpha + b * n_sample * n_ctype;
    double *beta0_b = beta0 + b * n_gene * n_ctype;
    double *phi0_b = phi0 + b * n_gene * n_ctype;
    double *beta_b = beta + b * n_coef * n_gene * n_ctype;
    double *phi_b = phi + b * n_disp * n_gene * n_ctype;
    uv *E_t = E + t_id * n_sample * n_gene;

    if( *norm_method == 2 ) // global
      {
      int dimg[2] = { n_sample * n_ctype, n_gene };
      malm11( dimg, Yg, w_b, Eg + t_id * n_sample * n_ctype * n_gene,
        alpha_b, beta0_b, phi0_b, NULL,
        iopt11_,dopt11_ );
      }
    else if( *norm_method == 1 ) // per cell type
      {
      for(int k = 0; k < n_ctype; k++ )
        {
        uv *Yk = Y + k * n_sample * n_gene;
        const double *w_bk = w_b + k * n_sample;
        double *alpha_bk = alpha_b + k * n_sample;
        double *beta0_bk = beta0_b + k * n_gene;
        double *phi0_bk = phi0_b + k * n_gene;

        malm11( dim, Yk, w_bk, E_t,
          alpha_bk, beta0_bk, phi0_bk, NULL,
          iopt11_, dopt11_
          );
        }
      }
    else
      ;  // no normalization

    if(n_coef > 0 )
      for(int k = 0; k < n_ctype; k++ )
        {
        uv *Yk = Y + k * n_sample * n_gene;
        const double *w_bk = w_b + k * n_sample;
        double *alpha_bk = alpha_b + k * n_sample;

        double *beta_bk = beta_b + k * n_coef * n_gene;
        double *phi_bk = phi_b + k * n_disp * n_gene;

        malm(dim, Yk, w_bk, Xt, Gi, E_t,
          alpha_bk, beta_bk, phi_bk, NULL,
          iopt_, dopt_, NULL);

        }

    #pragma omp critical
      {
      n_done++;
      if(iopt_[1]==1 && (n_done % chunk == 0))
        fprintf(stderr,"\rbootstrap completed: %d (%.1f%%)",
          n_done, (double)(n_done)*100/n_boot);
      }
    } // bootstrap

  if(iopt_[1]==1) fputc('\n',stderr);

  if( *norm_method == 2 )
    {
    for(int k = 1; k < n_ctype; k++ )
      for(int j = 0; j < n_gene; j++ )
        beta0[j + k*n_gene] = beta0[j];
    free(Yg);
    free(Eg);
    }
  free(w);
  free(E);
  return;
}
